### PR TITLE
Add infinite retry, fix monitoring

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?

Closes https://github.com/elastic/elastic-agent-shipper/issues/262

This PR adds infinite retry functionality, and completely refactors the elasticsearch health monitor. In detail

- If the `max_retries` is set to `-1`, this enables "infinite" retry, which just sets the ES MaxRetries to `MaxInt`.
- Because many common error conditions will now retry instead of trickling down to `OnFailure`, the monitor instead watches the raw HTTP return codes coming from the elasticsearch transport, and checks the deltas.
- The monitor will mark the ES output as unhealthy if zero successful HTTP codes are reported in a given period, and > 0 error conditions are reported across the reporting period

## Why is it important?

We want infinite retry.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.md` or `CHANGELOG-developer.md`.

## How to test this PR locally

- Pull down and build
- Run either standalone or with agent
- Set `max_retries` to `-1`:
```
outputs:
  default:
    log_level: debug
    type: elasticsearch
    hosts: [https://127.0.0.1:9200]
    username: "elastic"
    password: "changeme"
    bulk_max_size: 4096
    max_retries: -1
    ssl:
      verification_mode: none
    shipper:
      log_level: debug
      enabled: true
      bulk_max_size: 2048
```

- Ensure that events send properly
- Break Elasticsearch
- Run again, ensure that the shipper reports a degraded state

### CODE REVIEW

**Review by leehinman — APPROVED** [review]
LGTM

File: output/elasticsearch/watcher.go [hunk 1]
```diff
@@ -25,67 +27,136 @@ type WatchReporter func(state WatchState, msg string)
...
-	hw.lastSuccess = time.Now()
 }
 
-// Watch starts a blocking loop and will report if there has been a failure for longer than a given period
+// Watch starts a blocking loop and will report if there has been a failure for longer than a given period.
+// This is a blocking call
 func (hw *ESHealthWatcher) Watch(ctx context.Context) {
+
+	// count of failures from the es transport
+	lastFail := 0
```
Comment (faec) [inline output/elasticsearch/watcher.go:83]: Can we change some of these names to be more precise (e.g. it's kind of confusing when `lastFail` is an integer and `lastFailure` is a timestamp). Suggestions:
- `lastFail` -> `transportFailCount`
- `lastResponseFailRange` -> `statusFailCount`
- `lastResponseSuccessfulRange` -> `statusSuccessCount`

File: output/elasticsearch/watcher.go [hunk 2]
```diff
@@ -25,67 +27,136 @@ type WatchReporter func(state WatchState, msg string)
...
-		hw.mut.Lock()
+			// these return code ranges should probably be configurable?
+			success := returnRangeDelta(200, 299, metrics.Responses)
+			fails := returnRangeDelta(400, 599, metrics.Responses)
+
+			// calculate deltas between last run
+			respFailDelta := fails - lastResponseFailRange
+			respSuccessDelta := success - lastResponseSuccessfulRange
+			transportFailDelta := metrics.Failures - lastFail
+			cbFails := hw.callbackFailCount - lastCBFail
```
Comment (faec) [inline output/elasticsearch/watcher.go:113]: `callbackFailCount` should probably be checked via an atomic call

File: output/elasticsearch/watcher.go [hunk 3]
```diff
@@ -25,67 +27,136 @@ type WatchReporter func(state WatchState, msg string)
...
 func (hw *ESHealthWatcher) Watch(ctx context.Context) {
+
+	// count of failures from the es transport
+	lastFail := 0
+	// count of HTTP response codes in the fail range
+	lastResponseFailRange := 0
+	// count of HTTP response codes in the success range
+	lastResponseSuccessfulRange := 0
+	// is the watcher reported degraded?
+	didFail := false
```
Comment (faec) [inline output/elasticsearch/watcher.go:89]: it looks like `didFail` is all local now, but it still appears as a field in `ESHealthWatcher`, can that field be removed now?

**Review by faec — APPROVED** [review]
Looks good, a few suggested cleanups

**Pull request merged**